File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {

        unordered_map<char, int> letters;
        int start = 0; 
        int best = 0; 

        for (int end = 0; end < s.size(); end++){
            
            letters[s[end]]++;

            while(start < end && letters[s[end]] > 1){
                letters[s[start]]--;
                start++; 
            }

            best = max(best, end - start + 1);
        }

        return best; 
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        start = 0 
        end = 0

        letters = set()
        best = 0

        while end < len(s):

            while start < end and s[end] in letters: 
                letters.remove(s[start])
                start += 1
                

            letters.add(s[end])
            best = max(len(letters), best)
            end += 1

        return best

        